### PR TITLE
Add AdminController Test Suite

### LINKED ISSUE #174: Write admin controller test suite
Assertions we should test:
* /email:
    * Gets all the emails of stored users
* /bonus:
    * Updates points and activity to the users in the bonus request
    * Does not update points and activity to the users who aren't in the bonus request
*  /attendance (already implemented):
    * Logs activity, attendance, and points for users who have not attended
    * Does not log activity, attendance, and points for users who already attended
    * Logs proper activity and point rewards for staff attendance

### PR DESCRIPTION
**Pull request by StormFireFox1**
Closes #174.

### CODE REVIEW

**Comment by sumeet-bansal** [comment]
Heads up, you need to rebase (merge conflict).

**Comment by StormFireFox1** [comment]
Just did a squash on top of a rebase, since the merge conflict broke the lockfile pretty bad.

I'll keep this as a draft for another day or two, since I might be able to come up with more tests for the suite.

File: tests/admin.test.ts
Comment (sumeet-bansal) [inline tests/admin.test.ts:127]: Nit: rename to `admin`.

File: tests/admin.test.ts
Comment (sumeet-bansal) [inline tests/admin.test.ts:135]: Instead of sorting the arrays, there's an array equality check we can use. See the other tests for examples.

File: tests/admin.test.ts
Comment (sumeet-bansal) [inline tests/admin.test.ts:138]: I don't understand the point of this test. If the email doesn't exist, it wouldn't be returned because it doesn't exist. I'd understand if the endpoint returned some filtered list of addresses but that's not the case.

File: tests/admin.test.ts
Comment (sumeet-bansal) [inline tests/admin.test.ts:159]: Nit: rename to `admin`.

File: tests/admin.test.ts
Comment (sumeet-bansal) [inline tests/admin.test.ts:171]:
```suggestion
    const bonus = {
      description: 'Test addition of bonus points',
      users: emails,
      points: 200,
    };
    const createBonusResponse = await ControllerFactory.admin(conn).addBonus({ bonus }, admin);
    const getUserResponse = await ControllerFactory.user(conn).getUser({ uuid: users[0].uuid }, admin);
```

File: tests/admin.test.ts
Comment (sumeet-bansal) [inline tests/admin.test.ts:173]: If we're granting bonuses to multiple accounts, let's verify them all.

File: tests/admin.test.ts
Comment (sumeet-bansal) [inline tests/admin.test.ts:177]: This can be combined with the above test: a couple users get the bonus, a couple don't.

File: tests/admin.test.ts
Comment (shravanhariharan2) [reply tests/admin.test.ts:138]: Agreed, no filtering is happening or is planned, so I don't think this test is necessary

File: tests/admin.test.ts
Comment (shravanhariharan2) [inline tests/admin.test.ts:135]: nit: you can inline `adminController` since it's only used once:
```ts
const getAllEmailsResponse = await ControllerFactory.admin(conn).getAllEmails(admin);
```

File: tests/admin.test.ts
Comment (shravanhariharan2) [inline tests/admin.test.ts:171]: nit: with multiple assertions, we should assert right after getting a response (see Sumeet's explanation [here](https://github.com/acmucsd/membership-portal/pull/181#discussion_r626136760)), i.e.
```ts
const createBonusResponse = ...
expect(createBonusResponse.emails)...

const getUserResponse = ...
expect (getUserResponse.user)...
```

File: tests/admin.test.ts
Comment (StormFireFox1) [reply tests/admin.test.ts]: Resolved in 11c362c.

File: tests/admin.test.ts
Comment (StormFireFox1) [reply tests/admin.test.ts]: Resolved in fdbbd08.

File: tests/admin.test.ts
Comment (StormFireFox1) [reply tests/admin.test.ts]: Removed in 3a27e7f.

File: tests/admin.test.ts
Comment (StormFireFox1) [reply tests/admin.test.ts]: Resolved in 11c362c.

File: tests/admin.test.ts
Comment (StormFireFox1) [reply tests/admin.test.ts]: Resolved in 5fc6a9a.

File: tests/admin.test.ts
Comment (StormFireFox1) [reply tests/admin.test.ts]: Resolved in 5fe3084.

File: tests/admin.test.ts
Comment (StormFireFox1) [reply tests/admin.test.ts]: Resolved in 1e37dd2.

File: tests/admin.test.ts
Comment (StormFireFox1) [reply tests/admin.test.ts]: Resolved in 5fc6a9a. There are a few instances I inlined as well.

File: tests/admin.test.ts
Comment (StormFireFox1) [reply tests/admin.test.ts]: Resolved in 5ad4381.

Some other tests are required rearragements in the same file; I arranged those as well.

File: tests/admin.test.ts
Comment (shravanhariharan2) [inline tests/admin.test.ts:24]: nit: this should remain as `proxyUser`. The reason we have it named `proxyUser` here, but not elsewhere, is because this user is submitting attendance on behalf of the other users, and is hence a "proxy" (see definition 3 of [proxy definition](https://www.merriam-webster.com/dictionary/proxy)). We originally had it named as `onBehalfOfUser` but it didn't flow off the tongue as easily

File: tests/admin.test.ts
Comment (shravanhariharan2) [inline tests/admin.test.ts:130]: nit: rename `adminUser` to `admin` (for consistency with other tests + "user" is redundant since `admin` implies that the object is a user)

File: tests/admin.test.ts
Comment (shravanhariharan2) [inline tests/admin.test.ts:138]: nit: the assertions are flipped here between expected and actual. Every other test has:
```ts
expect(actual).toEqual(expected)
```  
but this one's:
```ts
expect(expected).toEqual(actual)
```

File: tests/admin.test.ts
Comment (shravanhariharan2) [inline tests/admin.test.ts:147]: nit: I believe you can do:
```ts
const [...users, extraneousUser] = UserFactory.create(5);
```
to declare both variables at once. I think this way of declaration makes the relationship between users and the extraneous user clearer than having them be declared separately

Comment (shravanhariharan2) [reply tests/admin.test.ts:147]: also `adminUser` -> `admin`

File: tests/admin.test.ts
Comment (shravanhariharan2) [reply tests/admin.test.ts:138]: also nit: inline `adminController`

**Review by shravanhariharan2 — COMMENTED** [review]
Small nitpicks on nomenclature/style for consistency, otherwise tests themselves LGTM

File: tests/admin.test.ts
Comment (sumeet-bansal) [reply tests/admin.test.ts:24]: I think `admin`'s fine since it's an admin account.

File: tests/admin.test.ts
Comment (sumeet-bansal) [reply tests/admin.test.ts:147]:
```typescript
const [...usersGettingBonus, userNotGettingBonus] = UserFactory.create(5);
```
or something similar.

File: tests/admin.test.ts
Comment (StormFireFox1) [reply tests/admin.test.ts]: I think `admin` would make the most sense, since the bonus route can only be used by an `admin` user anyway, so it also suggests proper flow for the endpoint as well.

Resolved in `df1a901`.

File: tests/admin.test.ts
Comment (StormFireFox1) [reply tests/admin.test.ts]: Resolved in `df1a901`.

File: tests/admin.test.ts
Comment (StormFireFox1) [reply tests/admin.test.ts]: Resolved in `aca6574`.

File: tests/admin.test.ts
Comment (StormFireFox1) [reply tests/admin.test.ts]: Resolved in `2afef57`.

**Review by shravanhariharan2 — APPROVED** [review]

File: tests/admin.test.ts
Comment (sumeet-bansal) [inline tests/admin.test.ts:80]: Can we verify this attendance is what we expect?

File: tests/admin.test.ts
Comment (sumeet-bansal) [inline tests/admin.test.ts:83]: Same here, let's verify the activity objects.

File: tests/admin.test.ts
Comment (sumeet-bansal) [inline tests/admin.test.ts:157]: Nit: let's declare this as late as possible, i.e. right before the loop.

File: tests/admin.test.ts
Comment (sumeet-bansal) [inline tests/admin.test.ts:165]: Let's also check that the users' activities are correct.

**Review by sumeet-bansal — COMMENTED** [review]
LGTM, just a couple assertions to add.

File: tests/admin.test.ts
Comment (StormFireFox1) [reply tests/admin.test.ts:80]: Resolved in 2c591a5.

File: tests/admin.test.ts
Comment (StormFireFox1) [reply tests/admin.test.ts]: Resolved in 2c591a5.

File: tests/admin.test.ts
Comment (StormFireFox1) [reply tests/admin.test.ts:168]: Resolved in 6688e90.

File: tests/admin.test.ts
Comment (StormFireFox1) [reply tests/admin.test.ts:85]: Resolved in 6688e90.

File: tests/admin.test.ts
Comment (sumeet-bansal) [inline tests/admin.test.ts:83]: What's this assertion for? This doesn't seem useful and can be removed.

**Review by sumeet-bansal — APPROVED** [review]

File: tests/admin.test.ts
Comment (StormFireFox1) [reply tests/admin.test.ts:83]: As per the previous review, this verifies the attendance response matches what we'd expect to receive when attending a specific event.

Comment (sumeet-bansal) [reply tests/admin.test.ts:83]: Ah right. LGTM but one of the tests is failing.

**Comment by StormFireFox1** [comment]
Due to a bug in the factory code causing events to randomly have start times before account creation times, it's possible for non-deterministic test suite failures because of account creation occurring after event attendance.

This bug is (supposedly) solved by PR #181 including the `fakeOngoingEvent` method. This PR is blocked until that one is merged.

**Comment by shravanhariharan2** [comment]
image

Here's an example - the event attendance timestamp occurs before the user creation since presumably, the event is randomly created in the past and has its timestamp set to a random time between start and end date. See https://github.com/acmucsd/membership-portal/blob/master/tests/data/PortalState.ts#L88

**Comment by sumeet-bansal** [comment]
@StormFireFox1 That PR hasn't been updated in a month and has changes requested, so let's just copy those factory changes over to this PR and merge this. I don't think there shouldn't be any merge conflicts, but even if there are, #181 can handle them.

**Comment by sumeet-bansal** [comment]
Actually @StormFireFox1 this probably affects a bunch of other tests too. We could just set the timestamp of the account creation activity to like a month ago to ensure that activity always happens first.

**Comment by StormFireFox1** [comment]
I figured yesterday explicitly setting account creation earlier is safer. I'll readjust all this tonight.

**Pull request merged**
